### PR TITLE
Add `rejection_reasons_type` column to `application choices`

### PR DESCRIPTION
**Pull request by steventux**
## Context

We're changing the format of structured reasons for rejection, the JSON we store on `ApplicationChoice` will also differ.

## Changes proposed in this pull request

- Adds a `rejection_reasons_type` column to `application choices`, with the default value of `reasons_for_rejection`, this maps to the model we use to hydrate the JSON.

## Guidance to review

The default value is intended to be constantized into the class name which can hydrate the `ApplicationChoice#structured_rejection_reasons` JSON value.
We can update this default value to the new type `rejection_reasons` when we release the redesign feature.
This should save us from backfilling and writing to the field for legacy reasons.

## Link to Trello card

https://trello.com/c/IP91rE34/4830-sr4r-redesign-add-sr4r-type-field-to-applicationchoices

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Review by JR-G — APPROVED** [review]

File: db/migrate/20220310110609_add_rejection_reasons_type_to_application_choices.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class AddRejectionReasonsTypeToApplicationChoices < ActiveRecord::Migration[6.1]
+  def up
+    add_column :application_choices, :rejection_reasons_type, :string, null: false, default: 'reasons_for_rejection'
```
Comment (cwrw) [inline db/migrate/20220310110609_add_rejection_reasons_type_to_application_choices.rb:3]: We currently have 2 types of reasons for rejection: structured and free text, do we want to differentiate those?

**Review by mnacos — COMMENTED** [review]
The only issue with a database migration setting a default field is that it may take a long time for large tables (locking them for certain updates in the process). We only have ~130k application choices so we can probably get away with this here.

Adding a new column with null values is a lot faster in postgres, so another way to go about this would be to add a nullable new column, then add a data migration setting values and eventually a schema migration setting a default/adding a not null constraint.

See https://wework.github.io/data/2015/11/05/add-columns-with-default-values-to-large-tables-in-rails-postgres/ -- not a recommendation, just an indication of what they experienced with ~2 million records.

**Comment by mnacos** [comment]
> See https://wework.github.io/data/2015/11/05/add-columns-with-default-values-to-large-tables-in-rails-postgres/ -- not a recommendation, just an indication of what they experienced with ~2 million records.

Did some more research and this may not be a problem any more. We are running Postgres 11, which includes an optimisation just for this: https://www.depesz.com/2018/04/04/waiting-for-postgresql-11-fast-alter-table-add-column-with-a-non-null-default/

**Comment by steventux** [comment]
Thanks @mnacos I'm gonna amend this migration to remove the default / not null options as we will likely use this field to denote structured AND flat rejection reasons so the usual add column -> write to column -> backfill approach is needed.

File: db/migrate/20220310110609_add_rejection_reasons_type_to_application_choices.rb (hunk 1 above)
Comment (duncanjbrown) [reply db/migrate/20220310110609_add_rejection_reasons_type_to_application_choices.rb:3]: Want to send this one to BigQuery too @claire-hughez?

**Comment by steventux** [comment]
Closing as it's easier to make another migration without the default value / null constraint

**Pull request closed without merging**